src/main/java/org/project/springweb/service/shoppingcart/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,113 @@
+        shoppingCart.setUser(user);
+        shoppingCartRepository.save(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto getShoppingCartForUser(Long userId) {
+        return shoppingCartMapper.toDto(findShoppingCartByUserId(userId));
+    }
+
+    @Override
+    public ShoppingCartResponseDto addCartItem(CreateCartItemRequestDto requestDto, Long userId) {
+        Book book = bookRepository.findById(requestDto.getBookId()).orElseThrow(
+                () -> new EntityNotFoundException("Cannot find book with id "
+                        + requestDto.getBookId())
+        );
+        ShoppingCart shoppingCart = findShoppingCartByUserId(userId);
+        Optional<CartItem> existingCartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> item.getBook().getId()
+                        .equals(book.getId()))
+                .findFirst();
+        CartItem cartItem;
+        if (existingCartItem.isPresent()) {
+            cartItem = existingCartItem.get();
+            UpdateCartItemRequestDto updateRequestDto = cartItemMapper
+                    .toUpdateCartItemRequestDto(requestDto);
+            updateCartItem(cartItem.getId(), updateRequestDto, userId);
+        } else {
+            cartItem = cartItemMapper.toEntity(requestDto);
+            cartItem.setShoppingCart(shoppingCart);
+            shoppingCart.getCartItems().add(cartItem);
+        }
+        cartItem.setBook(book);
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartResponseDto updateCartItem(Long cartId,
+                                                  UpdateCartItemRequestDto requestDto,
+                                                  Long userId) {
+        CartItem cartItem = findCartItemById(cartId);
+        ShoppingCart shoppingCart = checkIfCartItemBelongsToUser(userId, cartItem, cartId);
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteCartItem(Long cartId, Long userId) {
+        CartItem cartItem = findCartItemById(cartId);
+        checkIfCartItemBelongsToUser(userId, cartItem, cartId);
+        cartItemRepository.delete(cartItem);
+    }
+
+    private ShoppingCart findShoppingCartByUserId(Long userId) {
+        return shoppingCartRepository.findByUserId(userId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find shopping cart by userId "
+                        + userId)
+        );
+    }
+
+    private CartItem findCartItemById(Long cartId) {
+        return cartItemRepository.findById(cartId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find cart item by id "
+                        + cartId)
+        );
+    }
+
+    private ShoppingCart checkIfCartItemBelongsToUser(Long userId,
```

Comment:
If you create shop cart when user registers, I don't think that you need this checking
You can find shop cart by userId where is needed